src/vs/editor/contrib/anchorSelect/anchorSelect.ts

```diff
@@ -0,0 +1,172 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import 'vs/css!./anchorSelect';
+import { EditorAction, ServicesAccessor, registerEditorAction, registerEditorContribution } from 'vs/editor/browser/editorExtensions';
+import { localize } from 'vs/nls';
+import { EditorContextKeys } from 'vs/editor/common/editorContextKeys';
+import { Selection } from 'vs/editor/common/core/selection';
+import { KeyMod, KeyCode, KeyChord } from 'vs/base/common/keyCodes';
+import { KeybindingWeight } from 'vs/platform/keybinding/common/keybindingsRegistry';
+import { ICodeEditor } from 'vs/editor/browser/editorBrowser';
+import { RawContextKey, IContextKeyService, IContextKey } from 'vs/platform/contextkey/common/contextkey';
+import { IEditorContribution } from 'vs/editor/common/editorCommon';
+import { TrackedRangeStickiness } from 'vs/editor/common/model';
+import { MarkdownString } from 'vs/base/common/htmlContent';
+
+export const SelectionAnchorSet = new RawContextKey('selectionAnchorSet', false);
+
+class SelectionAnchorController implements IEditorContribution {
+
+	public static readonly ID = 'editor.contrib.selectionAnchorController';
+
+	static get(editor: ICodeEditor): SelectionAnchorController {
+		return editor.getContribution<SelectionAnchorController>(SelectionAnchorController.ID);
+	}
+
+	private decorationId: string | undefined;
+	private selectionAnchorSetContextKey: IContextKey<boolean>;
+
+	constructor(
+		private editor: ICodeEditor,
+		@IContextKeyService contextKeyService: IContextKeyService
+	) {
+		this.selectionAnchorSetContextKey = SelectionAnchorSet.bindTo(contextKeyService);
```

Comment:
I think this should also listen to `editor.onDidChangeModel` and reset the context key if the editor gets a new model set.